contracts/cw-core/src/contract.rs

```diff
@@ -312,35 +321,47 @@ pub fn execute_update_proposal_modules(
     env: Env,
     sender: Addr,
     to_add: Vec<ModuleInstantiateInfo>,
-    to_remove: Vec<String>,
+    to_disable: Vec<String>,
 ) -> Result<Response, ContractError> {
     if env.contract.address != sender {
         return Err(ContractError::Unauthorized {});
     }
 
-    for addr in to_remove {
+    let disable_count = to_disable.len() as u64;
+    for addr in to_disable {
         let addr = deps.api.addr_validate(&addr)?;
-        PROPOSAL_MODULES.remove(deps.storage, addr);
+        let mut module = PROPOSAL_MODULES
+            .load(deps.storage, addr.clone())
+            .map_err(|_| ContractError::ProposalModuleDoesNotExist {
+                address: addr.clone(),
+            })?;
+
+        if module.status == ProposalModuleStatus::Disabled {
+            return Err(ContractError::ModuleAlreadyDisabled {
+                address: module.address,
+            });
+        }
+
+        module.status = ProposalModuleStatus::Disabled {};
+        PROPOSAL_MODULES.save(deps.storage, addr, &module)?;
     }
 
+    // If disabling this module will cause there to be no active modules, return error.
+    // We don't check the active count before disabling because there may erroneously be
+    // modules in to_disable which are already disabled.
+    ACTIVE_PROPOSAL_MODULE_COUNT.update(deps.storage, |count| {
+        if count <= disable_count && to_add.is_empty() {
```

Comment:
```suggestion
        if count < disable_count && to_add.is_empty() {
```

In the `count == disable_count` case I _think_ we're actually OK here as we check that we will be adding at least one module with the `to_add.is_empty()` part of the condition. Otherwise I think we prevent disabling all modules and adding a new one here which is probably a pretty frequent operation. Might be nice to add a test for this.

Reply:
if we just do `count < disable_count && to_add.is_empty()`, we miss the case where `count == disable_count && to_add.is_empty()`, meaning we could disable all modules and not add any. The case where we disable all modules and add a new one will not pass this condition (because to_add.is_empty() will be false). I believe [this test should cover that scenario](https://github.com/DA0-DA0/dao-contracts/blob/19f06a42b55f9e118f8533a19c1d00df4d69e82c/contracts/cw-core/src/tests.rs#L431) (the (1,1) case)

Reply:
Oh, you're so right. Thank for the detailed response!